### ISSUE TITLE
Connecting to digital ocean managed database

### ISSUE THREAD

**Issue by ajaymore** [graphile/postgraphile#1455]
### Summary

I am getting below error

Error: self signed certificate in certificate chain
    at TLSSocket.onConnectSecure (_tls_wrap.js:1473:34)
    at TLSSocket.emit (events.js:311:20)
    at TLSSocket._finishInit (_tls_wrap.js:916:8)
    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:686:12)

Connection details

npx postgraphile -c "postgres://doadmin:secret@xxx-postgres-do-user-4864288-0.db.ondigitalocean.com:25060/apu-prod?ssl=1&sslrootcert=ca-certificate.crt" -s 'my_db' --retry-on-init-fail --watch --enhance-graphiql --dynamic-json

Tried few more combinations
npx postgraphile -c "postgres://doadmin:secret@xxx-postgres-do-user-4864288-0.db.ondigitalocean.com:25060/apu-prod?ssl=1&sslmode=require&sslrootcert=ca-certificate.crt" -s 'my_db' --retry-on-init-fail --watch --enhance-graphiql --dynamic-json

npx postgraphile -c "postgres://doadmin:secret@xxx-postgres-do-user-4864288-0.db.ondigitalocean.com:25060/apu-prod?sslmode=prefer" -s 'my_db' --retry-on-init-fail --watch --enhance-graphiql --dynamic-json

Below is connection string copied from digital ocean console.
postgresql://doadmin:secret@xxx-postgres-do-user-4864288-0.db.ondigitalocean.com:25060/apu-prod?sslmode=require

Am I missing something while running postgraphile?

**Comment by benjie** [comment]
Try giving an absolute path to the cert rather than relative

**Comment by benjie** [comment]
You can use sslmode=no-verify if you’re not concerned about verification

**Comment by ajaymore** [comment]
> Try giving an absolute path to the cert rather than relative

Tried that as well. 

```
npx postgraphile -c 'postgres://secret@parallelearning-postgres-do-user-4864288-0.db.ondigitalocean.com:25060/apu-prod?sslmode=require&sslrootcert=/Users/mystical/Downloads/ca-certificate.crt' -s 'my_db' --retry-on-init-fail --watch --enhance-graphiql --dynamic-json
```

sslmode=no-verify does not work as digital ocean managed database strictly enforces SSL

**Comment by benjie** [comment]
@ajaymore Try and get the connection running with the `pg` (node-postgres) module. If you can do that, try and get it running with the `pg` module using a `connectionString`. If that works then it should work in PostGraphile.

**Comment by benjie** [comment]
(I suspect the certificate you're passing doesn't have the content required to make the connection authenticated. But I don't know what to advise there. Here's one that works with RDS (note it's a certificate chain): https://github.com/graphile/starter/blob/main/data/amazon-rds-ca-cert.pem)

**Comment by ajaymore** [comment]
> @ajaymore Try and get the connection running with the `pg` (node-postgres) module. If you can do that, try and get it running with the `pg` module using a `connectionString`. If that works then it should work in PostGraphile.

It works with below configuration
```
const { Client } = require('pg');

const client = new Client({
  user: 'doadmin',
  host: 'parallelearning-postgres-do-user-4864288-0.db.ondigitalocean.com',
  database: 'shiparc',
  password: 'sdkcb7gd8wbzdssb',
  port: 25060,
  ssl: {
    rejectUnauthorized: false,
  },
});

client
  .connect()
  .then((info) => console.log(info))
  .catch((err) => console.log('errror', err));

Not sure how to convert this to connection string.
```

**Comment by benjie** [comment]
`?sslmode=no-verify` _should_ do the same assuming you're using this version of `pg-connection-string`:

https://github.com/brianc/node-postgres/blob/4b229275cfe41ca17b7d69bd39f91ada0068a5d0/packages/pg-connection-string/index.js#L96

Maybe you need to update `pg` and `pg-connection-string`?

**Comment by benjie** [comment]
In Graphile Starter I had to use yarn resolutions to force the use of the latest pg-connection-string: https://github.com/graphile/starter/blob/563b74f716cddbc6b915b79d7272a78720d5b42c/package.json#L83

**Comment by benjie** [comment]
@ajaymore what solved the issue in the end?

**Comment by ajaymore** [comment]
It didn't work for me, documentation for Digital Ocean managed database itself is quite sparse and also quite old.
https://github.com/digitalocean/databases/tree/master/examples/node-contacts

It is working with self hosted postgres so continuing with that.

**Comment by thomashermine** [comment]
If you are using [Postgraphile as a library](https://www.graphile.org/postgraphile/usage-library/), you can pass any options to PG in the first parameter. This solved it for me: 

```
const postgraphileServer = postgraphile({
    connectionString: POSTGIS_CS
     // Needed for DigitalOcean Postgres Managed Database ⤵️
    ssl: {
      rejectUnauthorized: false,
    },
  }, SCHMEMA, POSTGRAPHILE_OPTIONS)
```

**Comment by axhl** [comment]
There has been a lot of discussion in the Discord about using PostGraphile with DigitalOcean's DBaaS. Here are some quick summary notes that may be beneficial for complete beginners and which would have resolved many of the reported issues, including my own. 

**TL;DR: proceed with caution. The service isn't very mature yet, the docs are not good, and superuser access is not allowed.**

### Use Postgraphile as a Library – avoid the CLI

To successfully connect, follow this https://github.com/graphile/postgraphile/issues/1455#issuecomment-854420464. If you can figure out how to connect via the CLI, please comment, folks are currently terribly confused. 

### Keep your certificate easy to access

If you are using Digital Ocean's App Platform, don't bother with their environment variables functionality until you have PostGraphile successfully running to minimise confounding factors when debugging. 

### Remember to whitelist your Postgraphile IP

Under "Settings" in the control panel for your DO DB, you'll be able to add it under "Trusted Sources". 

### Remove `watchPg: true` 

Digital Ocean's managed Postgres service [does not allow a user to access the superuser role](https://docs.digitalocean.com/products/databases/postgresql/). This means you cannot create event triggers, and therefore can't use `watchPg`. To update your GraphQL API when your database schema changes, you'll have to relaunch your app. Bad DX.

**Comment by roelandmoors** [comment]
I can use the library like this:

```javascript
const dbCertPath = path.resolve("./ca-certificate.crt");
if (process.env.DATABASE_CA_CERT) {
  fs.writeFileSync(dbCertPath, process.env.DATABASE_CA_CERT);
}

const connString = process.env.DATABASE_URL + "&sslrootcert=" + dbCertPath;

app.use(
  postgraphile(
    connString,
    "app_public",
    production ? postgraphileOptionsProd : postgraphileOptionsDev
  )
);
```

I have 2 env variables:
DATABASE_URL: ${your_db_name.DATABASE_URL}
DATABASE_CA_CERT: ${your_db_name.CA_CERT}

**Comment by ajay-paramlogic** [comment]
Thank you all for your responses, they have helped a lot. This is what has worked for me.

```
postgraphile(
    {
      user: 'doadmin',
      host: 'xxxxxx-do-user-xxxxx-0.b.db.ondigitalocean.com',
      database: 'xxxxx',
      password: 'xxxxxxxx',
      port: 25060,
      ssl: {
        rejectUnauthorized: false,
      },
    },
    'public',
    {
      watchPg: false,
      graphiql: true,
      enhanceGraphiql: true,
      retryOnInitFail: true,
    }
  )

```

**Comment by jschatz1** [comment]
I've solved and understand this issue after researching for a while.  🥳 

Digitalocean managed databases work fine, there are no problems.  However there isn't a ton of documentation if you don't know what you're doing. 

You will see a lot of comments telling you to turn off rejecting unauthorized with `NODE_TLS_REJECT_UNAUTHORIZED` and other various methods, and while that will work, it defeats the purpose. Do not do this. You have everything you need to make this work the right way. 

DigitalOcean gives you a [CA file](https://www.ssl.com/faqs/what-is-a-certificate-authority/). 

DigitalOcean also does self sign their certificates. If you have the CA file then everything will be ok. 

You do not get the private key or a cert file from DO, since this would defeat the purpose of asymmetric encryption. So all you have to do is download the ca file from your managed database database section. 

<img width="568" alt="image" src="https://user-images.githubusercontent.com/322665/175824151-7b095029-2b01-4ae0-bfcc-e840a8340949.png">

You should be connecting via the VPC network, not the public network. In order to do that you need to make sure that your resources are all in the same VPC network. For me this was NYC1. 

<img width="584" alt="image" src="https://user-images.githubusercontent.com/322665/175824224-8c159d1f-12b4-4446-a42c-5050a80fc4d1.png">

For me I'm running a Kubernetes cluster, not a droplet, but it's the same method, and I need to make sure that my nodes can connect to my managed db so the managed db and the nodes are all in NYC1. 

I load the CA file via env vars, but you could store it in a file and load it, so I do this: 

```javascript
  "production": {
    "dialect" : "postgres",
    "dialectOptions":{
      "ssl": {
        "ca": process.env.DB_CERT
      }
    },
    "database": process.env.DB_NAME,
    "port": process.env.DB_PORT,
    "host": process.env.DB_HOST,
    "password": process.env.DB_PASSWORD,
    "sslmode": process.env.DB_SSLMODE,
    "username": process.env.DB_USERNAME,
```

this is my config for sequelize but the issue we are all facing is the same, because you need to pass the underlying connection:

```javascript
      "ssl": {
        "ca": process.env.DB_CERT
      }
```

You could also do a `fs.readFileSync('yourcafile')`

Hope this helps.

**Comment by ajay-paramlogic** [comment]
This worked for me. Thank you all for help.

```js
import express from "express";
import { postgraphile } from "postgraphile";
import fs from "fs";

const middleware = postgraphile(
  {
    database: process.env.DB_NAME,
    port: process.env.DB_PORT,
    host: process.env.DB_HOST,
    password: process.env.DB_PASSWORD,
    user: process.env.DB_USERNAME,
    sslmode: "require",
    ssl: {
      ca: fs.readFileSync("ca-certificate.crt"),
    },
  },
  "public",
  {
    watchPg: false,
    graphiql: true,
    enhanceGraphiql: true,
    retryOnInitFail: true,
  }
);

const app = express();
app.use(middleware);

const server = app.listen(3000, () => {
  const address = server.address();
  if (typeof address !== "string") {
    const href = `http://localhost:${address.port}${"/graphiql"}`;
    console.log(`PostGraphiQL available at ${href} 🚀`);
  } else {
    console.log(`PostGraphile listening on ${address} 🚀`);
  }
});
```
